_posts/2016/2016-11-21-multitenancy-and-elasticsearch.md

```diff
@@ -0,0 +1,127 @@
+
+The first option is an easy and straightforward concept.
+It provides solid data isolation from a storage perspective and you can still query across multiple indexes if you need to. This solution is sufficient for most cases.
+
+Each index comes with some overhead and this can easily add up.
+By introducing a new index per tenant, you are multiplying your index by the number of tenants `T` you have.
+If you then introduce localization for `L` languages and need to create a new index for that as well.
+In each of those, you might have `T` different types, which you might (or cannot) keep in the same index as well.
+Considering that, you end up with `T * L * T` indices for one installation.
+As an example, for 10,000 tenants, 10 types and 10 supported languages, you will need 1 million indices.
+
+### Shared Index
+
+To keep the amount of indices under control, you can share the same index for all tenants.
+This would reduce the number of indices to 100,000 for the example mentioned above.
+
+Elasticsearch provides some features that help in dealing with multitenancy in a single index.
+
+#### Bool / Filtered Query
+
+A simple way to implement this is by wrapping queries into a [bool query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-bool-query.html) and filter by tenant.
+This is very similar to what you might be used to with other databases and corresponding frameworks.
+
+{% highlight javascript %}
+POST /products/_search
+{
+  "query" : {
+    "bool" : {
+      "filter" : {
+        "term" : { "tenantId" : "12345" }
+      },
+      // ...
+    }
+  }
+}
+{% endhighlight %}
+
+This is a possible solution, but you would probably like to have some kind of framework support for that, otherwise you forget to wrap a query somewhere and end up exposing data across tenants.
+
+#### Index Alias
+
+Another solution nifty feature you can use are [filtered aliases](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-aliases.html#filtered).
+
+They basically enable you to do the same thing as with filtered queries by introducing an alias for every tenant.
+This way you an deal with tenancy in a more transparent way - you don't need wrap every query yourself, you let elasticsearch do it for you.
+
+{% highlight javascript %}
+POST /_aliases
+{
+  "actions" : [
+    {
+      "add" : {
+         "index" : "products",
+         "alias" : "products_12345",
+         "filter" : { "term" : { "tenantId" : "12345" } }
+      }
+    }
+  ]
+}
+{% endhighlight %}
+
+It is also more secure when it comes to data separation because you will need to specify the index/alias you want to query and forgetting about the tenant will lead to the index not being found instead of retrieving the results for all tenants.
+
+By delegating the this responsibility to elasticsearch for the most part, you can get rid of some complexity in your code, but you need to take care of creating and deleting these aliases.
```

Comment:
By delegating the this responsibility to elasticsearch for the most part, you can get rid of some complexity in your code, but you need to take care of creating and deleting these aliases. >
By delegating this responsibility to Elasticsearch for the most part, you can get rid of some complexity in your code, but you need to take care of creating and deleting these aliases.